src/Compilers/CSharp/Portable/Symbols/Source/SourceMemberContainerSymbol.cs

```diff
@@ -3455,6 +3552,15 @@ private void CheckForStructBadInitializers(DeclaredMembersAndInitializersBuilder
             }
         }
 
+        private void AddSynthesizedSimpleProgramEntryPointIfNecessary(MembersAndInitializersBuilder builder, DeclaredMembersAndInitializers declaredMembersAndInitializers, BindingDiagnosticBag diagnostics)
+        {
+            var simpleProgramEntryPoints = GetSimpleProgramEntryPoints();
+            foreach (var member in simpleProgramEntryPoints)
+            {
+                builder.AddNonTypeMember(member, declaredMembersAndInitializers);
```

Comment:
>AddNonTypeMember(member, declaredMembersAndInitializers) [](http://example.com/codeflow?start=24&length=56)

Do we have a race condition where external callers of `GetSimpleProgramEntryPoints()` might get back an entry point that hasn't been added to `MembersAndInitializers` yet? If so, what are the implications?

Reply:
I don't think that is possible. `GetSimpleProgramEntryPoints` is atomic. It will return exactly one set of entry points, then we'll incorporate them into `MembersAndInitializers`. 
Different threads will see the same set of entry points, and there is not other way to get entry point symbols than through `GetSimpleProgramEntryPoints`.

Does that make sense, or do you still think there might be a way?